lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,456 @@
+        @event_deque = []
+      end
+
+      # The number of bundled elements in the repeated field.
+      # @return [Numeric]
+      def element_count
+        sum = 0
+        @in_deque.each { |elts| sum += elts.count }
+        sum
+      end
+
+      # The size of the request in bytes of the bundled field elements.
+      # @return [Numeric]
+      def request_bytesize
+        sum = 0
+        @in_deque.each do |elts|
+          elts.each do |e|
+            sum += e.to_s.bytesize
+          end
+        end
+        sum
+      end
+
+      # Call the task's api_call.
+      #
+      # The task's func will be called with the bundling requests function.
+      def run
+        return if @in_deque.count == 0
+        request = @bundling_request
+        request.send(@bundled_field).clear
+        request.send(@bundled_field).concat(@in_deque.flatten)
+        if !@subresponse_field.nil?
+          run_with_subresponses(request, @kwargs)
+        else
+          run_with_no_subresponse(request, @kwargs)
+        end
+      end
+
+      # Disable this since the api_call may raise any type of exception.
+      # rubocop:disable Lint/RescueException
+
+      # Helper for #run to run the api call with no subresponses.
+      #
+      # @param request [Object] the request to pass as the arg to
+      #     the api_call.
+      # @param kwargs [Hash] the keyword arguments passed to the api_call.
+      def run_with_no_subresponse(request, kwargs)
+        response = @api_call.call(request, **kwargs)
+        @event_deque.each do |event|
+          event.result = response
+          event.set
+        end
+      rescue Exception => err
+        @event_deque.each do |event|
+          event.result = err
+          event.set
+        end
+      ensure
+        @in_deque.clear
+        @event_deque.clear
+      end
+
+      # Helper for #run to run the api call with subresponses.
+      #
+      # @param request [Object] the request to pass as the arg to
+      #     the api_call.
+      # @param subresponse_field subresponse_field.
+      # @param kwargs [Hash] the keyword arguments passed to the api_call.
+      def run_with_subresponses(request, kwargs)
+        response = @api_call.call(request, **kwargs)
+        in_sizes_sum = 0
+        @in_deque.each { |elts| in_sizes_sum += elts.count }
+        all_subresponses = response.send(@subresponse_field.to_s)
+        if all_subresponses.count != in_sizes_sum
+          # TODO: Implement a logging class to handle this.
+          # warn DEMUX_WARNING
+          @event_deque.each do |event|
+            event.result = response
+            event.set
+          end
+        else
+          start = 0
+          @in_deque.zip(@event_deque).each do |i, event|
+            response_copy = response.dup
+            subresponses = all_subresponses[start, i.count]
+            response_copy.send(@subresponse_field).clear
+            response_copy.send(@subresponse_field).concat(subresponses)
```

Comment:
Since protobuf needs types to be enforced this will not work. Probably response_copy[@subresponse_field] = []
response_copy[@subresponse_field] += subresponses
https://developers.google.com/protocol-buffers/docs/reference/ruby-generated#repeated-fields